File: backend/src/localy/api/pool/endpoints.py
"""
Device pooling API routes (Phase 3).

    GET  /pool/status          — current pool + coordinator state
    POST /pool/join            — add a worker (manual host:port)
    POST /pool/leave           — remove a worker
    GET  /pool/plan/{model}    — shard plan for a model across the pool
    GET  /pool/fit/{model}     — pool-fit advisor (does it fit with current pool?)
    POST /pool/load            — load a model split across the pool
    POST /pool/unload          — stop pooled inference
"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from localy.core.config import Settings, get_settings
from localy.core.dependencies import verify_api_key
from localy.schemas.pool import (
    JoinRequest,
    LeaveRequest,
    PoolLoadRequest,
    PoolNodeResponse,
    PoolStatusResponse,
    ShardPlanResponse,
)
from localy.services.pool_service import get_pool_service

pool_router = APIRouter(prefix="/pool", tags=["Pooling (Phase 3)"])


@pool_router.get("/status", response_model=PoolStatusResponse, dependencies=[Depends(verify_api_key)])
async def pool_status(settings: Settings = Depends(get_settings)) -> PoolStatusResponse:
    """Return the current pool membership and coordinator state."""
    return PoolStatusResponse(**get_pool_service(settings).status())


@pool_router.post("/join", response_model=PoolNodeResponse, dependencies=[Depends(verify_api_key)])
async def pool_join(req: JoinRequest, settings: Settings = Depends(get_settings)) -> PoolNodeResponse:
    """Add a remote worker to the pool."""
    budget_bytes = req.budget_mib * 1024 * 1024 if req.budget_mib else None
    node = get_pool_service(settings).join(
        host=req.host, port=req.port, label=req.label, budget_bytes=budget_bytes
    )
    return PoolNodeResponse(
        node_id=node.node_id,
        address=node.address,
        is_local=node.is_local,
        label=node.label,
        budget_gb=round(node.budget_gb, 2),
    )


@pool_router.post("/leave", dependencies=[Depends(verify_api_key)])
async def pool_leave(req: LeaveRequest, settings: Settings = Depends(get_settings)) -> dict[str, bool]:
    """Remove a worker from the pool."""
    return {"removed": get_pool_service(settings).leave(req.node_id)}


@pool_router.get("/discover", dependencies=[Depends(verify_api_key)])
async def pool_discover(
    auto_join: bool = False, settings: Settings = Depends(get_settings)
) -> list[dict]:
    """List pool workers advertised on the LAN via mDNS (optionally auto-join them)."""
    return get_pool_service(settings).discover(auto_join=auto_join)


@pool_router.get(
    "/plan/{model_id:path}",
    response_model=ShardPlanResponse,
    dependencies=[Depends(verify_api_key)],
)
async def pool_plan(model_id: str, settings: Settings = Depends(get_settings)) -> ShardPlanResponse:
    """Compute how a model would be split across the current pool."""
    plan = get_pool_service(settings).plan_for_model(model_id)
    return ShardPlanResponse(**plan.to_dict())


@pool_router.get(
    "/fit/{model_id:path}",
    response_model=ShardPlanResponse,
    dependencies=[Depends(verify_api_key)],
)
async def pool_fit(model_id: str, settings: Settings = Depends(get_settings)) -> ShardPlanResponse:
    """Pool-fit advisor: same as plan, framed as a fit check for the UI."""
    plan = get_pool_service(settings).plan_for_model(model_id)
    return ShardPlanResponse(**plan.to_dict())


@pool_router.post("/load", response_model=ShardPlanResponse, dependencies=[Depends(verify_api_key)])
async def pool_load(req: PoolLoadRequest, settings: Settings = Depends(get_settings)) -> ShardPlanResponse:
    """Load a model split across the pool (spawns the coordinator llama-server)."""
    plan = get_pool_service(settings).load_pooled(req.model, n_ctx=req.ctx)
    return ShardPlanResponse(**plan.to_dict())


@pool_router.post("/unload", dependencies=[Depends(verify_api_key)])
async def pool_unload(settings: Settings = Depends(get_settings)) -> dict[str, str]:
    """Stop pooled inference."""
    get_pool_service(settings).unload_pooled()
    return {"status": "stopped"}

File: desktop/src/api/endpoints.ts
// API endpoint mappings
import { apiClient, API_BASE_URL } from "./client";
import {
  HardwareReport,
  FitAssessment,
  RegistryModel,
  BenchmarkResult,
  PoolStatus,
  ShardPlan,
  DiscoveredWorker,
  AccessInfo,
} from "./types";

export const api = {
  /**
   * Get basic health status
   */
  async getHealth(): Promise<{ status: string }> {
    return apiClient.get<{ status: string }>("/health");
  },

  /**
   * Get readiness status (active model info, etc.)
   */
  async getReadiness(): Promise<{ status: string; model_loaded: boolean; active_model: string | null }> {
    return apiClient.get<{ status: string; model_loaded: boolean; active_model: string | null }>("/ready");
  },

  /**
   * Get detailed hardware capability report
   */
  async getHardwareReport(): Promise<HardwareReport> {
    return apiClient.get<HardwareReport>("/system/hardware");
  },

  /**
   * Get detailed fit assessment for a model
   */
  async getModelFit(modelId: string, context?: number): Promise<FitAssessment> {
    const query = context ? `?context=${context}` : "";
    return apiClient.get<FitAssessment>(`/system/hardware/fit/${encodeURIComponent(modelId)}${query}`);
  },

  /**
   * Get list of all registered models annotated with download status and fit assessment.
   * Variants are fetched live from Hugging Face (cached).
   */
  async getModels(): Promise<RegistryModel[]> {
    return apiClient.get<RegistryModel[]>("/system/models");
  },

  /** Extract text from a document (PDF/text/code/md) to use as chat context. */
  async extractDocument(
    file: File
  ): Promise<{ filename: string; text: string; chars: number; truncated: boolean; error?: string }> {
    const fd = new FormData();
    fd.append("file", file);
    const res = await fetch(`${API_BASE_URL}/system/extract`, { method: "POST", body: fd });
    if (!res.ok) throw new Error(`Extraction failed (HTTP ${res.status})`);
    return res.json();
  },

  /**
   * Search Hugging Face for GGUF models to add to the catalog.
   */
  async searchCatalog(query: string): Promise<{ id: string; downloads: number; likes: number }[]> {
    return apiClient.get<{ id: string; downloads: number; likes: number }[]>(
      `/system/catalog/search?q=${encodeURIComponent(query)}`
    );
  },

  /**
   * Add a Hugging Face GGUF repo to the catalog (all its variants become available).
   */
  async addCatalogModel(repoId: string): Promise<{ id?: string; variants?: number; error?: string }> {
    return apiClient.post<{ id?: string; variants?: number; error?: string }>(
      "/system/catalog/add",
      { repo_id: repoId }
    );
  },

  /**
   * Trigger a standardized benchmark run on a model
   */
  async runBenchmark(modelId: string, iterations: number = 3): Promise<BenchmarkResult> {
    return apiClient.post<BenchmarkResult>("/system/benchmark", {
      model: modelId,
      iterations,
    });
  },

  /**
   * Get historical benchmark results
   */
  async getBenchmarkHistory(): Promise<BenchmarkResult[]> {
    return apiClient.get<BenchmarkResult[]>("/system/benchmark/history");
  },

  /**
   * Delete a locally downloaded model variant
   */
  async deleteModel(modelId: string): Promise<{ status: string }> {
    return apiClient.delete<{ status: string }>("/api/delete", {
      name: modelId,
    });
  },

  // --- Background downloads (server-side; survive tab switches) ---

  /** Start (or resume) a background download. Returns immediately. */
  async startDownload(model: string): Promise<any> {
    return apiClient.post<any>("/system/downloads/start", { model });
  },

  /** Progress of all downloads this session. */
  async getDownloads(): Promise<
    { model_id: string; status: string; completed: number; total: number; speed_mbps: number; error?: string | null }[]
  > {
    return apiClient.get("/system/downloads");
  },

  /** Cancel a background download (partial kept for resume). */
  async cancelDownload(model: string): Promise<any> {
    return apiClient.post<any>("/system/downloads/cancel", { model });
  },

  // --- Pooling (Phase 3) ---

  /** Current pool membership + coordinator state. */
  async getPoolStatus(): Promise<PoolStatus> {
    return apiClient.get<PoolStatus>("/pool/status");
  },

  /** Workers advertised on the LAN via mDNS (optionally auto-join). */
  async discoverPool(autoJoin: boolean = false): Promise<DiscoveredWorker[]> {
    return apiClient.get<DiscoveredWorker[]>(`/pool/discover?auto_join=${autoJoin}`);
  },

  /** Add a worker by address. Pass budgetGb from discovery so the planner uses
   *  the worker's REAL memory (otherwise the backend falls back to a guess). */
  async joinPool(host: string, port: number, label: string = "", budgetGb?: number | null): Promise<any> {
    const body: Record<string, unknown> = { host, port, label };
    if (budgetGb && budgetGb > 0) body.budget_mib = Math.round(budgetGb * 1024);
    return apiClient.post<any>("/pool/join", body);
  },

  /** Remove a worker from the pool. */
  async leavePool(nodeId: string): Promise<{ removed: boolean }> {
    return apiClient.post<{ removed: boolean }>("/pool/leave", { node_id: nodeId });
  },

  /** Pool-fit advisor: does this model fit across the current pool?
   *  May fetch model metadata from HF, so allow more than the default timeout. */
  async poolFit(modelId: string): Promise<ShardPlan> {
    return apiClient.get<ShardPlan>(`/pool/fit/${encodeURIComponent(modelId)}`, { timeoutMs: 60000 });
  },

  /** Load a model split across the pool (spawns the coordinator).
   *  This can take minutes (streaming weights to remote workers over WiFi), so
   *  the client timeout is disabled — the backend enforces its own 900s cap. */
  async loadPooled(model: string, ctx: number = 4096): Promise<ShardPlan> {
    return apiClient.post<ShardPlan>("/pool/load", { model, ctx }, { timeoutMs: 0 });
  },

  /** Stop pooled inference. */
  async unloadPooled(): Promise<{ status: string }> {
    return apiClient.post<{ status: string }>("/pool/unload", {}, { timeoutMs: 30000 });
  },

  /** Share THIS device as a worker (start rpc-server + advertise). */
  async startWorker(): Promise<{ running: boolean; address?: string }> {
    return apiClient.post<{ running: boolean; address?: string }>("/pool/worker/start", {}, { timeoutMs: 60000 });
  },

  /** Stop sharing this device. */
  async stopWorker(): Promise<{ running: boolean }> {
    return apiClient.post<{ running: boolean }>("/pool/worker/stop", {});
  },

  // --- API access (keys + internet tunnel) ---

  /** LAN/local URLs, keys, and tunnel status for the API Access panel. */
  async getAccess(): Promise<AccessInfo> {
    return apiClient.get<AccessInfo>("/system/access");
  },

  /** Generate a new API key. The full `key` is returned once — copy it now. */
  async createKey(label: string): Promise<{ id: string; key: string; label: string }> {
    return apiClient.post<{ id: string; key: string; label: string }>("/system/keys", { label });
  },

  /** Revoke an API key by id. */
  async revokeKey(id: string): Promise<{ revoked: boolean }> {
    return apiClient.delete<{ revoked: boolean }>(`/system/keys/${id}`);
  },

  /** Start the Cloudflare internet tunnel; returns the public URL. */
  async startTunnel(): Promise<{ running: boolean; url: string | null }> {
    return apiClient.post<{ running: boolean; url: string | null }>("/system/tunnel/start", {});
  },

  /** Stop the internet tunnel. */
  async stopTunnel(): Promise<{ running: boolean }> {
    return apiClient.post<{ running: boolean }>("/system/tunnel/stop", {});
  },
};
